src/magnifier.c

```diff
@@ -0,0 +1,296 @@
+	double dst_x = 0;
+	double dst_y = 0;
+
+	/* Ensure everything is kept within output boundaries */
+	if (x < 0) {
+		cropped_width += x;
+		dst_x = x * -1;
+		x = 0;
+	}
+	if (y < 0) {
+		cropped_height += y;
+		dst_y = y * -1;
+		y = 0;
+	}
+	cropped_width = MIN(cropped_width, (double)output_buffer->width - x);
+	cropped_height = MIN(cropped_height, (double)output_buffer->height - y);
+
+	/* (Re)create the temporary buffer if required */
+	if (tmp_buffer && (tmp_buffer->width != width || tmp_buffer->height != height)) {
+		wlr_log(WLR_ERROR, "tmp buffer size changed, dropping");
+		assert(tmp_texture);
+		wlr_texture_destroy(tmp_texture);
+		wlr_buffer_drop(tmp_buffer);
+		tmp_buffer = NULL;
+		tmp_texture = NULL;
+	}
+	if (!tmp_buffer) {
+		tmp_buffer = wlr_allocator_create_buffer(
+			server->allocator, width, height, &wlr_drm_format);
+	}
+	if (!tmp_buffer) {
+		wlr_log(WLR_ERROR, "Failed to allocate temporary magnifier buffer");
+		return;
+	}
+
+	/* Extract source region into temporary buffer */
+
+	struct wlr_render_pass *tmp_render_pass = wlr_renderer_begin_buffer_pass(
+		server->renderer, tmp_buffer, NULL);
+
+	/* FIXME, try to re-use the existing output texture instead */
+	wlr_buffer_lock(output_buffer);
+	struct wlr_texture *output_texture = wlr_texture_from_buffer(
+		server->renderer, output_buffer);
+	assert(output_texture);
+
+	struct wlr_render_texture_options opts = {
+		.texture = output_texture,
+		.src_box = (struct wlr_fbox) {
+			x, y, cropped_width, cropped_height },
+		.dst_box = (struct wlr_box) {
+			dst_x, dst_y, cropped_width, cropped_height },
+		.alpha = NULL,
+	};
+	wlr_render_pass_add_texture(tmp_render_pass, &opts);
+	if (!wlr_render_pass_submit(tmp_render_pass)) {
+		wlr_log(WLR_ERROR, "Failed to extract magnifier source region");
+		wlr_texture_destroy(output_texture);
+		goto cleanup;
+	}
+	wlr_texture_destroy(output_texture);
+
+	/* Render to the output buffer itself */
+	tmp_render_pass = wlr_renderer_begin_buffer_pass(
+		server->renderer, output_buffer, NULL);
+
+	/* Borders */
+	if (fullscreen) {
+		border_box.x = 0;
+		border_box.y = 0;
+		border_box.width = width;
+		border_box.height = height;
+	} else {
+		border_box.x = ox - (width / 2 + theme->mag_border_width);
+		border_box.y = oy - (height / 2 + theme->mag_border_width);
+		border_box.width = (width + theme->mag_border_width * 2);
+		border_box.height = (height + theme->mag_border_width * 2);
+		struct wlr_render_rect_options bg_opts = {
+			.box = border_box,
+			.color = (struct wlr_render_color) {
+				.r = theme->mag_border_color[0],
+				.g = theme->mag_border_color[1],
+				.b = theme->mag_border_color[2],
+				.a = theme->mag_border_color[3]
+			},
+			.clip = NULL,
+		};
+		wlr_render_pass_add_rect(tmp_render_pass, &bg_opts);
+	}
+
+	/* Paste the magnified result back into the output buffer */
+	if (!tmp_texture) {
+		tmp_texture = wlr_texture_from_buffer(server->renderer, tmp_buffer);
+		assert(tmp_texture);
+	}
```

Comment:
AFAIR `wlr_texture_from_buffer()` may fail, thus this should `return;` instead of the assert() and the whole block should be moved to the top after the `wlr_log(WLR_ERROR, "Failed to allocate temporary magnifier buffer")` branch.

Reply:
Done